## Cosmological Calculations (astropy.cosmology)

### Introduction

The astropy.cosmology subpackage contains classes for representing cosmologies, and utility functions for calculating commonly used quantities that depend on a cosmological model. This includes distances, ages and lookback times corresponding to a measured redshift or the transverse separation corresponding to a measured angular separation.

### Getting Started

Cosmological quantities are calculated using methods of a Cosmology object. For example, to calculate the Hubble constant at z=0 (i.e., H0), and the number of transverse proper kpc corresponding to an arcminute at z=3:

In [28]:
from astropy.cosmology import WMAP9 as cosmo
from astropy.cosmology import WMAP7   # WMAP 7-year cosmology
from astropy.cosmology import FlatLambdaCDM
import astropy.units as u
from astropy.cosmology import FlatwCDM
from astropy.cosmology import Planck13, z_at_value

In [3]:
cosmo.H(0)  

<Quantity 69.32 km / (Mpc s)>

In [4]:
cosmo.kpc_proper_per_arcmin(3)  

<Quantity 472.97709236 kpc / arcmin>

Here WMAP9 is a built-in object describing a cosmology with the parameters from the 9-year WMAP results. Several other built-in cosmologies are also available, see Built-in Cosmologies. The available methods of the cosmology object are listed in the methods summary for the FLRW class. If you’re using IPython you can also use tab completion to print a list of the available methods. To do this, after importing the cosmology as in the above example, type cosmo. at the IPython prompt and then press the tab key.

All of these methods also accept an array of redshifts as input:

In [5]:
cosmo.comoving_distance([0.5, 1.0, 1.5])  

<Quantity [1916.06942039, 3363.0706321 , 4451.74754107] Mpc>

You can create your own arbitrary cosmology using one of the Cosmology classes:

In [13]:
cosmo = FlatLambdaCDM(H0=70, Om0=0.3)

In [8]:
cosmo

FlatLambdaCDM(H0=70 km / (Mpc s), Om0=0.3, Tcmb0=0 K, Neff=3.04, m_nu=None, Ob0=None)

The cosmology subpackage makes use of units, so in many cases returns values with units attached. Consult the documentation for that subpackage for more details, but briefly, to access the floating point or array values:

In [10]:
H0 = cosmo.H(0)

In [11]:
H0.value, H0.unit

(70.0, Unit("km / (Mpc s)"))

### Using astropy.cosmology

Most of the functionality is enabled by the FLRW object. This represents a homogeneous and isotropic cosmology (characterized by the Friedmann-Lemaitre-Robertson-Walker metric, named after the people who solved Einstein’s field equation for this special case). However, you can’t work with this class directly, as you must specify a dark energy model by using one of its subclasses instead, such as FlatLambdaCDM.

You can create a new FlatLambdaCDM object with arguments giving the Hubble parameter and omega matter (both at z=0):

In [12]:
from astropy.cosmology import FlatLambdaCDM

In [14]:
cosmo = FlatLambdaCDM(H0=70, Om0=0.3)

In [15]:
cosmo

FlatLambdaCDM(H0=70 km / (Mpc s), Om0=0.3, Tcmb0=0 K, Neff=3.04, m_nu=None, Ob0=None)

This can also be done more explicitly using units, which is recommended:

In [17]:
cosmo = FlatLambdaCDM(H0=70 * u.km / u.s / u.Mpc, Om0=0.3)

However, most of the parameters that accept units (H0, Tcmb0) have default units, so unit quantities do not have to be used. The exception are neutrino masses, where you must supply a units if you want massive neutrinos.

The pre-defined cosmologies described in the Getting Started section are instances of FlatLambdaCDM, and have the same methods. So we can find the luminosity distance to redshift 4 by:

In [18]:
cosmo.luminosity_distance(4)  

<Quantity 35851.83207232 Mpc>

or the age of the universe at z = 0:

In [19]:
cosmo.age(0)  

<Quantity 13.46698395 Gyr>

They also accept arrays of redshifts:

In [21]:
cosmo.age([0.5, 1, 1.5]).value  

array([8.42634602, 5.75164694, 4.20073193])

See the [FLRW](http://docs.astropy.org/en/stable/api/astropy.cosmology.FLRW.html#flrw) and [FlatLambdaCDM](http://docs.astropy.org/en/stable/api/astropy.cosmology.FlatLambdaCDM.html#astropy.cosmology.FlatLambdaCDM) object docstring for all the methods and attributes available. In addition to flat Universes, non-flat varieties are supported such as LambdaCDM. There are also a variety of standard cosmologies with the parameters already defined (see Built-in Cosmologies):

In [23]:
WMAP7.critical_density(0)  # critical density at z = 0  

<Quantity 9.30966846e-30 g / cm3>

You can see how the density parameters evolve with redshift as well:

In [24]:
WMAP7.Om([0, 1.0, 2.0]), WMAP7.Ode([0., 1.0, 2.0])  

(array([0.272     , 0.74898523, 0.90905237]),
 array([0.72791572, 0.25055061, 0.0901026 ]))

Note that these don’t quite add up to one even though WMAP7 assumes a flat Universe because photons and neutrinos are included. Also note that they are unitless and so are not Quantity objects.

Cosmological instances have an optional name attribute which can be used to describe the cosmology:

In [26]:
cosmo = FlatwCDM(name='SNLS3+WMAP7', H0=71.58, Om0=0.262, w0=-1.016)

In [27]:
cosmo

FlatwCDM(name="SNLS3+WMAP7", H0=71.6 km / (Mpc s), Om0=0.262, w0=-1.02, Tcmb0=0 K, Neff=3.04, m_nu=None, Ob0=None)

This is also an example with a different model for dark energy, a flat Universe with a constant dark energy equation of state, but not necessarily a cosmological constant. A variety of additional dark energy models are also supported – see Specifying a dark energy model.

A important point is that the cosmological parameters of each instance are immutable – that is, if you want to change, say, Om, you need to make a new instance of the class.

Finding the Redshift at a Given Value of a Cosmological Quantity
If you know a cosmological quantity and you want to know the redshift which it corresponds to, you can use z_at_value:

In [29]:
z_at_value(Planck13.age, 2 * u.Gyr)  

3.198120965652982

For some quantities there can be more than one redshift that satisfies a value. In this case you can use the zmin and zmax keywords to restrict the search range. See the z_at_value docstring for more detailed usage examples.

Built-in Cosmologies
A number of pre-loaded cosmologies are available from analyses using the WMAP and Planck satellite data. For example,

In [30]:
Planck13.lookback_time(2)  # lookback time in Gyr at z=2  

<Quantity 10.51184138 Gyr>

A full list of the pre-defined cosmologies is given by cosmology.parameters.available, and summarized below:

|Name	 |Source	                    |H0	    |Om	    |Flat|
| --- | --- | --- | --- | --- |
|WMAP5	 |Komatsu et al. 2009	        |70.2	|0.277	|Yes |
|WMAP7	 |Komatsu et al. 2011	        |70.4	|0.272	|Yes |
|WMAP9	 |Hinshaw et al. 2013	        |69.3	|0.287	|Yes |
|Planck13|Planck Collab 2013, Paper XVI	|67.8	|0.307	|Yes |

Currently, all are instances of FlatLambdaCDM. More details about exactly where each set of parameters come from are available in the docstring for each object:

In [31]:
print(WMAP7.__doc__)

WMAP7 instance of FlatLambdaCDM cosmology

(from Komatsu et al. 2011, ApJS, 192, 18, doi: 10.1088/0067-0049/192/2/18. Table 1 (WMAP + BAO + H0 ML).)


### Specifying a dark energy model

In addition to the standard FlatLambdaCDM model described above, a number of additional dark energy models are provided. FlatLambdaCDM and LambdaCDM assume that dark energy is a cosmological constant, and should be the most commonly used cases; the former assumes a flat Universe, the latter allows for spatial curvature. FlatwCDM and wCDM assume a constant dark energy equation of state parameterized by $w_{0}$. Two forms of a variable dark energy equation of state are provided: the simple first order linear expansion $w(z) = w_{0} + w_{z} z$ by w0wzCDM, as well as the common CPL form by w0waCDM: $w(z) = w_{0} + w_{a} (1 - a) = w_{0} + w_{a} z / (1 + z)$ and its generalization to include a pivot redshift by wpwaCDM: $w(z) = w_{p} + w_{a}(a_{p} - a)$.

Users can specify their own equation of state by sub-classing FLRW. See the provided subclasses for examples.

### Photons and Neutrinos

The cosmology classes include the contribution to the energy density from both photons and neutrinos. By default, the latter are assumed massless. The three parameters controlling the proporties of these species, which are arguments to the initializers of all the cosmological classes, are Tcmb0 (the temperature of the CMB at z=0), Neff, the effective number of neutrino species, and m_nu, the rest mass of the neutrino species. Tcmb0 and m_nu should be expressed as unit Quantities. All three have standard default values (2.725 K, 3.04, and 0 eV respectively; the reason that Neff is not 3 primarily has to do with a small bump in the neutrino energy spectrum due to electron-positron annihilation, but is also affected by weak interaction physics).

Massive neutrinos are treated using the approach described in the WMAP 7-year cosmology paper (Komatsu et al. 2011, ApJS, 192, 18, section 3.3). This is not the simple $\Omega_{\nu 0} h^2 = \sum_i m_{\nu\, i} / 93.04\,\mathrm{eV}$ approximation. Also note that the values of \Omega_{\nu}(z) include both the kinetic energy and the rest-mass energy components, and that the Planck13 cosmology includes a single species of neutrinos with non-zero mass (which is not included in $\Omega_{m0}$).

The contribution of photons and neutrinos to the total mass-energy density can be found as a function of redshift:

In [32]:
WMAP7.Ogamma0, WMAP7.Onu0  # Current epoch values  

(4.985868989189756e-05, 3.442275090310745e-05)

In [34]:
z = [0, 1.0, 2.0]

In [35]:
WMAP7.Ogamma(z), WMAP7.Onu(z)

(array([4.98586899e-05, 2.74583989e-04, 4.99898824e-04]),
 array([3.44227509e-05, 1.89574501e-04, 3.45133270e-04]))

If you want to exclude photons and neutrinos from your calculations, simply set Tcmb0 to 0:

In [47]:
from astropy.cosmology import FlatLambdaCDM
cos = FlatLambdaCDM(70.4 * u.km / u.s / u.Mpc, 0.272, Tcmb0 = 0.0 * u.K)

In [48]:
cos.Ogamma0, cos.Onu0

(0.0, 0.0)

Neutrinos can be removed (while leaving photons) by setting Neff to 0:

In [74]:
cos = FlatLambdaCDM(70.4, 0.272, Neff=0)

In [75]:
cos.Ogamma([0, 1, 2])  # Photons are still present   ?????????

array([0., 0., 0.])

In [76]:
cos.Onu([0, 1, 2])  # But not neutrinos

array([0., 0., 0.])

The number of neutrino species is assumed to be the floor of Neff, which in the default case is 3. Therefore, if non-zero neutrino masses are desired, then 3 masses should be provided. However, if only one value is provided, all the species are assumed to have the same mass. Neff is assumed to be shared equally between each species.

In [77]:
H0 = 70.4 * u.km / u.s / u.Mpc
m_nu = 0 * u.eV
cosmo = FlatLambdaCDM(H0, 0.272, m_nu=m_nu)

In [78]:
cosmo.has_massive_nu

False

In [79]:
print(cosmo.m_nu)    #Should be zeros

None


In [80]:
m_nu = [0.0, 0.05, 0.10] * u.eV
cosmo = FlatLambdaCDM(H0, 0.272, m_nu=m_nu)

In [81]:
cosmo.has_massive_nu

False

In [82]:
cosmo.m_nu     #????????? <Quantity [ 0.  , 0.05, 0.1 ] eV>

In [83]:
cosmo.Onu([0, 1.0, 15.0])       #?????? array([ 0.00326988,  0.00896783,  0.0125786 ])

array([0., 0., 0.])

In [84]:
cosmo.Onu(1) * cosmo.critical_density(1)   #??? <Quantity 2.444380380370406e-31 g / cm3>

<Quantity 0. g / cm3>

While these examples used FlatLambdaCDM, the above examples also apply for all of the other cosmology classes.

### For Developers: Using astropy.cosmology inside Astropy

If you are writing code for the Astropy core or an affiliated package, it’s often useful to assume a default cosmology, so that the exact cosmology doesn’t have to be specified every time a function or method is called. In this case it’s possible to specify a “default” cosmology.

You can set the default cosmology to a pre-defined value by using the “default_cosmology” option in the [cosmology.core] section of the configuration file (see Configuration system (astropy.config)). Alternatively, you can use the set function of default_cosmology to set a cosmology for the current Python session. If you haven’t set a default cosmology using one of the methods described above, then the cosmology module will default to using the 9-year WMAP parameters.

It is strongly recommended that you use the default cosmology through the default_cosmology science state object. An override option can then be provided using something like the following:

def myfunc(..., cosmo=None):
    from astropy.cosmology import default_cosmology

    if cosmo is None:
        cosmo = default_cosmology.get()

    ... your code here ...
This ensures that all code consistently uses the default cosmology unless explicitly overridden.

Note

In general it’s better to use an explicit cosmology (for example WMAP9.H(0) instead of cosmology.default_cosmology.get().H(0)). Use of the default cosmology should generally be reserved for code that will be included in the Astropy core or an affiliated package.

See Also
Hogg, “Distance measures in cosmology”, http://arxiv.org/abs/astroph/9905116
Linder, “Exploring the Expansion History of the Universe”, http://arxiv.org/abs/astro-ph/0208512
NASA’s Legacy Archive for Microwave Background Data Analysis, http://lambda.gsfc.nasa.gov/
Range of validity and reliability
The code in this sub-package is tested against several widely-used online cosmology calculators, and has been used to perform many calculations in refereed papers. You can check the range of redshifts over which the code is regularly tested in the module astropy.cosmology.tests.test_cosmology. If you find any bugs, please let us know by opening an issue at the github repository!

The built in cosmologies use the parameters as listed in the respective papers. These provide only a limited range of precision, and so you should not expect derived quantities to match beyond that precision. For example, the Planck 2013 results only provide the Hubble constant to 4 digits. Therefore, the Planck13 built-in cosmology should only be expected to match the age of the Universe quoted by the Planck team to 4 digits, although they provide 5 in the paper.

Reference/API
astropy.cosmology Module
astropy.cosmology contains classes and functions for cosmological distance measures and other cosmology-related calculations.

See the Astropy documentation for more detailed usage examples and references.